File: backend/src/main/java/it/unical/inf/ea/backend/data/services/interfaces/UserService.java
package it.unical.inf.ea.backend.data.services.interfaces;

import com.nimbusds.jose.JOSEException;
import it.unical.inf.ea.backend.data.entities.User;
import it.unical.inf.ea.backend.dto.UserDTO;
import it.unical.inf.ea.backend.dto.basics.UserBasicDTO;
import it.unical.inf.ea.backend.dto.enums.Provider;
import it.unical.inf.ea.backend.dto.enums.UserRole;
import jakarta.mail.MessagingException;
import jakarta.persistence.EntityNotFoundException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.transaction.Transactional;
import org.springframework.data.domain.Page;
import org.springframework.http.ResponseEntity;

import java.io.IOException;
import java.text.ParseException;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

public interface UserService {
    UserDTO createUser(User user);

    UserDTO partialUpdateUser(UUID id, Map<String, Object> updates) throws IllegalAccessException;

    UserBasicDTO updateMe(Map<String, Object> updates) throws IllegalAccessException;

    void deleteUser(UUID id);

    UserBasicDTO findUserById(UUID id);

    Optional<UserBasicDTO> findBasicByUsername(String username);

    Page<UserDTO> findAll(int page, int size, UserRole userRole, String username) throws IllegalAccessException;

    Optional<UserDTO> findByUsername(String username);

    List<UserDTO> getAllUser() throws IllegalAccessException;

    Map<String, String> googleAuth(String code) throws Exception;


    Map<String, String> authenticateUser(String username, String password, Provider provider) throws JOSEException;

    ResponseEntity<String> registerUser(String firstname, String lastname, String email, String password) throws MessagingException;

    ResponseEntity<String> sendVerificationEmail(String username) throws MessagingException;

    void rejectToken(HttpServletRequest request) throws ParseException;

    Map<String, String> refreshToken(String authorizationHeader, HttpServletResponse response) throws IOException, ParseException, JOSEException;

    void createUser(String lastname, String firstname, String email, String password) throws MessagingException;

    void activateUser(String token) throws ParseException, JOSEException;


    UserDTO changeRole(UUID userId, UserRole role) throws IllegalAccessException;

    UserDTO banUser(UUID userId);

    UserDTO unBanUser(UUID userId);


    void logout(HttpServletRequest request) throws ParseException, JOSEException;

    void changePassword(String authorizationHeader, String oldPassword, String newPassword) throws ParseException, JOSEException;

    void getNewPasswordByEmail(String token) throws ParseException, JOSEException, MessagingException;

    void resetPassword(String email) throws MessagingException;

    UserDTO getUserDTO() throws EntityNotFoundException;

    UserBasicDTO getUserBasicDTO() throws EntityNotFoundException;


}

File: frontend/app/src/main/java/com/android/frontend/service/UserService.kt
package com.android.frontend.service

import com.android.frontend.dto.UserDTO
import com.android.frontend.dto.basic.UserBasicDTO
import com.android.frontend.dto.UserUpdateRequest
import retrofit2.Call
import retrofit2.Response
import retrofit2.http.*

interface UserService {
    @POST("/api/v1/users/register")
    fun register(
        @Query("firstname") firstname: String,
        @Query("lastname") lastname: String,
        @Query("email") email: String,
        @Query("password") password: String
    ): Call<Void>

    @POST("/api/v1/users/login")
    fun login(
        @Query("username") username: String,
        @Query("password") password: String
    ): Call<Map<String, String>>

    @POST("/api/v1/users/changeRole/{userId}")
    suspend fun changeRole(
        @Header("Authorization") token: String,
        @Path("userId") userId: String,
        @Query("role") role: String
    ): Response<Void>

    @DELETE("/api/v1/users/deleteUser/{id}")
    suspend fun deleteUser(
        @Header("Authorization") token: String,
        @Path("id") userId: String
    ): Response<Void>

    @GET("/api/v1/users/refreshToken")
    fun refreshToken(
        @Header("Authorization") authorization: String
    ): Call<Map<String, String>>

    @GET("/api/v1/users/rejectToken")
    fun rejectToken(
        @Header("Authorization") authorization: String
    ): Call<Void>

    @GET("/api/v1/users/resetPassword")
    fun resetPassword(
        @Query("email") email: String
    ): Call<Void>

    @GET("/api/v1/users/me")
    fun userBasicDTOCall(
        @Header("Authorization") authorization: String
    ): Call<UserBasicDTO>

    @GET("/api/v1/users/retrieveUserProfile")
    fun userDTOCall(
        @Header("Authorization") authorization: String
    ): Call<UserDTO>

    @GET("/api/v1/users/getAllUsers")
    suspend fun getAllUsers(
        @Header("Authorization") token: String
    ): Response<List<UserDTO>>

    @GET("/api/v1/users/findUserById/{id}")
    fun findUserById(
        @Path("id") id: String
    ): Call<UserDTO>

    @GET("/api/v1/users/findByUsername")
    fun findByUsername(
        @Query("username") username: String
    ): Call<UserBasicDTO>

    @PATCH("/api/v1/updateUser/{id}")
    fun updateUser(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
        @Body patch: UserUpdateRequest
    ): Call<UserDTO>

    @DELETE("/api/v1/users/{id}")
    fun deleteUser(
        @Path("id") id: String
    ): Call<Void>

    @POST("/api/v1/users/changePassword")
    fun changePassword(
        @Header("Authorization") authorization: String,
        @Query("oldPassword") oldPpassword: String,
        @Query("newPassword") newPpassword: String
    ): Call<Void>

    @GET("/api/v1/users/getNewPassword")
    fun resetPasswordToken(
        @Query("token") token: String
    ): Call<Void>
}
